File: experiments/adrianco/experiment-13/runs/language=go_model=opus-4.8-fast_prompt=neutral/rep1/soccer/format.go
// format.go: human-readable rendering of matches, records, standings and player
// lists. The MCP tools return plain text (LLM-friendly), so these helpers turn
// the analysis structs from stats.go into the answer formats shown in the spec.
package soccer

import (
	"fmt"
	"strings"
)

// formatDate renders a match date as YYYY-MM-DD, or "(date unknown)".
func formatDate(m Match) string {
	if m.Date.IsZero() {
		return "(date unknown)"
	}
	return m.Date.Format("2006-01-02")
}

// matchContext renders the trailing "(Competition Round/Stage)" annotation.
func matchContext(m Match) string {
	parts := []string{m.Competition}
	if m.Round != "" {
		parts = append(parts, "Round "+m.Round)
	}
	if m.Stage != "" {
		parts = append(parts, m.Stage)
	}
	if m.Season != 0 && m.Round == "" && m.Stage == "" {
		parts = append(parts, fmt.Sprintf("%d", m.Season))
	}
	return "(" + strings.Join(parts, " ") + ")"
}

// FormatMatch renders a single match line, e.g.
// "2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Série A Round 22)".
func FormatMatch(m Match) string {
	return fmt.Sprintf("%s: %s %d-%d %s %s",
		formatDate(m), m.HomeTeam, m.HomeGoals, m.AwayGoals, m.AwayTeam, matchContext(m))
}

// FormatMatchList renders up to limit matches, with a trailing "(N more)" note.
func FormatMatchList(matches []Match, limit int) string {
	if len(matches) == 0 {
		return "No matches found."
	}
	var b strings.Builder
	shown := matches
	if limit > 0 && len(matches) > limit {
		shown = matches[:limit]
	}
	for _, m := range shown {
		fmt.Fprintf(&b, "- %s\n", FormatMatch(m))
	}
	if len(matches) > len(shown) {
		fmt.Fprintf(&b, "- ... (%d more matches in dataset)\n", len(matches)-len(shown))
	}
	return strings.TrimRight(b.String(), "\n")
}

// FormatRecord renders a Record as a multi-line block.
func FormatRecord(r Record, title string) string {
	return fmt.Sprintf("%s:\n- Matches: %d\n- Wins: %d, Draws: %d, Losses: %d\n"+
		"- Goals For: %d, Goals Against: %d (diff %+d)\n- Points: %d\n- Win rate: %.1f%%",
		title, r.Matches, r.Wins, r.Draws, r.Losses,
		r.GoalsFor, r.GoalsAgainst, r.GoalDiff(), r.Points(), r.WinRate()*100)
}

// FormatStandings renders a league table (top limit rows; 0 = all).
func FormatStandings(table []Record, title string, limit int) string {
	if len(table) == 0 {
		return "No standings available."
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%s:\n", title)
	rows := table
	if limit > 0 && len(rows) > limit {
		rows = rows[:limit]
	}
	for i, r := range rows {
		fmt.Fprintf(&b, "%2d. %s - %d pts (%dW, %dD, %dL), GF %d GA %d (%+d)\n",
			i+1, r.Team, r.Points(), r.Wins, r.Draws, r.Losses, r.GoalsFor, r.GoalsAgainst, r.GoalDiff())
	}
	return strings.TrimRight(b.String(), "\n")
}

// FormatPlayer renders a one-line player summary.
func FormatPlayer(p Player) string {
	club := p.Club
	if club == "" {
		club = "(no club)"
	}
	return fmt.Sprintf("%s - Overall: %d, Potential: %d, Position: %s, Age: %d, Nationality: %s, Club: %s",
		p.Name, p.Overall, p.Potential, dash(p.Position), p.Age, p.Nationality, club)
}

// FormatPlayerList renders up to limit players as a numbered list.
func FormatPlayerList(players []Player, limit int) string {
	if len(players) == 0 {
		return "No players found."
	}
	var b strings.Builder
	shown := players
	if limit > 0 && len(players) > limit {
		shown = players[:limit]
	}
	for i, p := range shown {
		fmt.Fprintf(&b, "%d. %s\n", i+1, FormatPlayer(p))
	}
	if len(players) > len(shown) {
		fmt.Fprintf(&b, "... (%d more players match)\n", len(players)-len(shown))
	}
	return strings.TrimRight(b.String(), "\n")
}

func dash(s string) string {
	if strings.TrimSpace(s) == "" {
		return "-"
	}
	return s
}

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=none/rep3/src/format.ts
/**
 * ============================================================================
 * Context: Brazilian Soccer MCP Server — Output Formatting
 * ----------------------------------------------------------------------------
 * Purpose : Render query results as compact, human-readable text matching the
 *           answer formats in the spec (match lists, team records, standings,
 *           player rankings). These strings are what the MCP tools return to
 *           the LLM, so they are kept terse and information-dense.
 * Consumers: server.ts.
 * ============================================================================
 */

import type { Match, Player, StandingRow, TeamRecord } from "./types.js";
import type { AggregateStats, HeadToHead, TeamRanking } from "./queries.js";

function score(m: Match): string {
  const h = m.homeGoals === null ? "?" : m.homeGoals;
  const a = m.awayGoals === null ? "?" : m.awayGoals;
  return `${h}-${a}`;
}

function context(m: Match): string {
  const bits = [m.competition];
  if (m.round) bits.push(`Round ${m.round}`);
  if (m.stage) bits.push(m.stage);
  return bits.join(", ");
}

/** Format a single match line: `2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Série A, Round 22)`. */
export function formatMatch(m: Match): string {
  const date = m.date ?? "date unknown";
  return `${date}: ${m.homeTeam} ${score(m)} ${m.awayTeam} (${context(m)})`;
}

export function formatMatchList(matches: Match[], limit = 25): string {
  if (matches.length === 0) return "No matches found.";
  const shown = matches.slice(0, limit).map((m) => `- ${formatMatch(m)}`);
  let out = shown.join("\n");
  if (matches.length > limit) {
    out += `\n- ... (${matches.length - limit} more match${matches.length - limit === 1 ? "" : "es"})`;
  }
  out = `Found ${matches.length} match${matches.length === 1 ? "" : "es"}:\n${out}`;
  return out;
}

export function formatTeamRecord(r: TeamRecord, heading?: string): string {
  const head = heading ?? `${r.team} record`;
  return [
    `${head}:`,
    `- Matches: ${r.matches}`,
    `- Wins: ${r.wins}, Draws: ${r.draws}, Losses: ${r.losses}`,
    `- Goals For: ${r.goalsFor}, Goals Against: ${r.goalsAgainst} (GD ${r.goalsFor - r.goalsAgainst})`,
    `- Points: ${r.points}`,
    `- Win rate: ${(r.winRate * 100).toFixed(1)}%`,
  ].join("\n");
}

export function formatHeadToHead(h: HeadToHead): string {
  const lines = [
    `Head-to-head: ${h.teamA} vs ${h.teamB} (${h.matches.length} meetings in dataset)`,
    `- ${h.teamA} wins: ${h.teamAWins}`,
    `- ${h.teamB} wins: ${h.teamBWins}`,
    `- Draws: ${h.draws}`,
    `- Goals: ${h.teamA} ${h.teamAGoals} — ${h.teamBGoals} ${h.teamB}`,
  ];
  if (h.matches.length > 0) {
    lines.push("", "Recent meetings:");
    for (const m of h.matches.slice(0, 10)) lines.push(`- ${formatMatch(m)}`);
  }
  return lines.join("\n");
}

export function formatPlayer(p: Player, index?: number): string {
  const prefix = index !== undefined ? `${index}. ` : "";
  const rating = p.overall ?? "?";
  const parts = [`${prefix}${p.name} — Overall: ${rating}`];
  if (p.position) parts.push(`Position: ${p.position}`);
  if (p.club) parts.push(`Club: ${p.club}`);
  if (p.nationality) parts.push(`Nationality: ${p.nationality}`);
  if (p.age !== null) parts.push(`Age: ${p.age}`);
  return parts.join(", ");
}

export function formatPlayerList(players: Player[], total?: number): string {
  if (players.length === 0) return "No players found.";
  const header = `Found ${total ?? players.length} player${(total ?? players.length) === 1 ? "" : "s"}${
    total && total > players.length ? ` (showing top ${players.length})` : ""
  }:`;
  const lines = players.map((p, i) => formatPlayer(p, i + 1));
  return [header, ...lines].join("\n");
}

export function formatStandings(rows: StandingRow[], title: string, limit = 30): string {
  if (rows.length === 0) return `No standings could be calculated for ${title}.`;
  const lines = rows.slice(0, limit).map((r) => {
    const marker = r.position === 1 ? " — Champion" : "";
    return `${r.position}. ${r.team} — ${r.points} pts (${r.wins}W ${r.draws}D ${r.losses}L, GF ${r.goalsFor} GA ${r.goalsAgainst}, GD ${r.goalDifference >= 0 ? "+" : ""}${r.goalDifference})${marker}`;
  });
  return [`${title} (calculated from matches):`, ...lines].join("\n");
}

export function formatAggregate(stats: AggregateStats, label: string): string {
  return [
    `${label}:`,
    `- Matches (with score): ${stats.matchesWithScore} / ${stats.matches}`,
    `- Total goals: ${stats.totalGoals}`,
    `- Average goals per match: ${stats.goalsPerMatch.toFixed(2)}`,
    `- Home win rate: ${(stats.homeWinRate * 100).toFixed(1)}%`,
    `- Away win rate: ${(stats.awayWinRate * 100).toFixed(1)}%`,
    `- Draw rate: ${(stats.drawRate * 100).toFixed(1)}%`,
  ].join("\n");
}

export function formatTeamRankings(rankings: TeamRanking[], label: string, asPercent = false): string {
  if (rankings.length === 0) return `No teams found for ${label}.`;
  const lines = rankings.map((r, i) => {
    const v = asPercent ? `${(r.value * 100).toFixed(1)}%` : `${r.value}`;
    return `${i + 1}. ${r.team} — ${v} (${r.record.wins}W ${r.record.draws}D ${r.record.losses}L over ${r.record.matches} matches)`;
  });
  return [`${label}:`, ...lines].join("\n");
}
